### PR TITLE
Make TryRun thread safe

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Changes
- Make `IImpelementationTask.TryRun` thread-safe
- Add `IImpelementationTask.IsIdle`

### Testing
- Update tests to exercise new behavior.

### CODE REVIEW

File: Source/ExecutionEngine/ImplementationTask.cs [hunk 1]
```diff
@@ -62,40 +72,59 @@ public class ImplementationTask : IImplementationTask {
...
   public IObservable<IVerificationStatus>? TryRun()
   {
     if (CacheStatus is Completed) {
       return null;
     }
 
-    var alreadyRunning = cancellationSource != null;
-    if (alreadyRunning) {
-      return null;
+    lock (myLock) {
```
Comment (MikaelMayer) [inline Source/ExecutionEngine/ImplementationTask.cs:89]: Shouldn't the lock also contain the previous if statement?
Two threads could arrive at this lock at the same time and a third one sets `CacheStatus` as `Completed` before it enters the lock.
Then there could be race conditions.

**Review by MikaelMayer — COMMENTED** [review]
Pending questions. Please explain me what you are achieving with enough detail so that I can understand your design.

File: Source/ExecutionEngine/ImplementationTask.cs [hunk 2]
```diff
@@ -62,40 +72,78 @@ public class ImplementationTask : IImplementationTask {
...
+
+      cancellationSource = new();
     }
 
-    var alreadyRunning = cancellationSource != null;
-    if (alreadyRunning) {
+    // We claimed the right to run. No other thread is running nor can they get to this point,
+    // so from this point we can read and write from and to the fields CacheStatus and status as if we're the only thread.
+
+    if (CacheStatus is Completed) {
```
Comment (cpitclaudel) [inline Source/ExecutionEngine/ImplementationTask.cs:120]: Could we do this part before creating the new `cancellationSource` above?

File: Source/ExecutionEngine/ImplementationTask.cs [hunk 3]
```diff
@@ -62,40 +72,78 @@ public class ImplementationTask : IImplementationTask {
...
     task.ContinueWith(r =>
     {
-      if (r.Exception != null) {
-        observableStatus.OnError(r.Exception);
-      } else {
-        observableStatus.OnCompleted();
+      // Lock so we may do operations after clearing cancellationSource,
+      // which releases our control over the field status.
+      lock (myLock) {
+        // Clear cancellationSource before calling status.OnCompleted, so ImplementationTask.IsIdle returns false
```
Comment (cpitclaudel) [inline Source/ExecutionEngine/ImplementationTask.cs:137]: `false` or `true`?

**Review by cpitclaudel — CHANGES_REQUESTED** [review]
I left a few comments.  Happy to re-review after that, I think this is pretty close to approval.

File: Source/Directory.Build.props [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
   <!-- Target framework and package configuration -->
   <PropertyGroup>
-    <Version>2.15.6</Version>
+    <Version>2.15.7</Version>
```
[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/Directory.Build.props:5]: Bumps the Boogie version number. I do this on most of my Boogie PRs so I'm then also able to use the Boogie changes in Dafny.

File: Source/ExecutionEngine/ImplementationTask.cs [hunk 5]
```diff
@@ -62,40 +72,78 @@ public class ImplementationTask : IImplementationTask {
   }
 
   private CancellationTokenSource? cancellationSource;
+  private ReplaySubject<IVerificationStatus>? status;
```
[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/ExecutionEngine/ImplementationTask.cs:75]: status always ends in either Completed or Stale, and CacheStatus reflects that final value. If no run has been done, so no status has been set, CacheStatus is Stale.

File: Source/ExecutionEngine/ImplementationTask.cs [hunk 6]
```diff
@@ -62,40 +72,78 @@ public class ImplementationTask : IImplementationTask {
...
+        return result;
+      }
+
+      cancellationSource = new();
     }
 
-    var alreadyRunning = cancellationSource != null;
-    if (alreadyRunning) {
+    // We claimed the right to run. No other thread is running nor can they get to this point,
+    // so from this point we can read and write from and to the fields CacheStatus and status as if we're the only thread.
```
[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/ExecutionEngine/ImplementationTask.cs]: The large lock makes it harder to refactor, but I've made an attempt :)

File: Source/ExecutionEngine/ImplementationTask.cs [hunk 7]
```diff
@@ -62,40 +72,93 @@ public class ImplementationTask : IImplementationTask {
...
+
+      if (cancellationSource.IsCancellationRequested) {
+        // Another thread is running but was cancelled,
+        // so we may immediately start a new run after the cancellation completes.
+        return QueueRun();
+      }
+
+      // Another thread is running and is not cancelled, so this run fails.
       return null;
     }
```
Comment (cpitclaudel) [inline Source/ExecutionEngine/ImplementationTask.cs:98]: Does the whole method now run inside the lock?

File: Source/ExecutionEngine/ImplementationTask.cs [hunk 8]
```diff
@@ -62,40 +72,93 @@ public class ImplementationTask : IImplementationTask {
...
+        }
       }
     }, TaskScheduler.Current);
-    return observableStatus;
+    return status;
+  }
+
+  private IObservable<IVerificationStatus>? QueueRun()
+  {
+
```
Comment (cpitclaudel) [inline Source/ExecutionEngine/ImplementationTask.cs:138]:
```suggestion
```

**Review by cpitclaudel — APPROVED** [review]
Looks good to me.  The one thing I'm not 100% clear on is whether there would be a way to unify the two cases (enqueuing to wait for the cancelled task to exit vs starting a task right away).  I'm also not 100% sure that I understand whether the `status` variable is sufficiently protected: is there a chance that the status can change while we hold the lock, and would that cause an issue?

File: Source/ExecutionEngine/ImplementationTask.cs (hunk 7 above)
Comment (keyboardDrummer) [reply Source/ExecutionEngine/ImplementationTask.cs:98]: Jep

**Comment by keyboardDrummer** [comment]
>  is there a chance that the status can change while we hold the lock, and would that cause an issue?

I don't think so, but it seems impossible to say for sure :-)

**Review by cpitclaudel — APPROVED** [review]

**Pull request merged**